### DISCUSSION TITLE
Intermittent Alarm 1 and noise on Limit switches

### DISCUSSION THREAD

**Discussion by Klippert** [grblHAL/core#568] [Q&A] [answered]
I am currently battling a problem with noise on my limit switches which in turn results in an Alarm 1, and i know this is not exactly a software issue, but i felt that it would still be a good place to document, as i didn't find much about it while searching the discussion forum. 

I started this documentation in the AutoSquaring conversation, but figure it'll be easier to find if it has it's own topic. 

I'm using a Teensy on Phils breakout board T41U5XBB (https://github.com/phil-barrett/grblHAL-teensy-4.x) witch has worked excellent for the last 3 years. 

Now i've rebuilt my CNC with new Z and X axis and many more improvements. 

Among other things, i've moved from mechanical limitswitches to optical ones. (https://www.aliexpress.com/item/1005002900573779.html) - which are supposed to be a clone of this one: https://www.ia.omron.com/product/item/2223/ 
![image](https://github.com/user-attachments/assets/6986812b-09a2-4641-a19c-d8331304123a)

My CNC has a 2.2kw china spindle with a equally chinese VFD. 

I've experienced the following:
Problem: Alarm 1 occurring when i turn on the spindle. Every time. 
Solution: Nothing - the problem disappeared after a day of rest. - now i can't reproduce that particular problem.

Problem: Alarm 1 when running a program of about 20minutes with 6 toolchanges. Alarm 1 would occour about 1-2 minutes into the program.
Solution: I replaced one limitswitch on one of the Y axis, and the problem went away. After this i tried running the program several times cutting air by running Z high - and everything worked. Tool-changes, probing etc.

Problem: I get intermittent Alarm 1 when i home the machine, move around by jogging - but strangely not when turning on and off the spindle. 
Solution: so far not found. 
I've done some measuring with my pocket scope, and found that the limit inputs on my breakout board are surprisingly noisy. 

Z Limit Signal: (that's a std. mechanical endstop like this one:
image)
![image](https://github.com/user-attachments/assets/5afa6868-b2a5-4bcb-a1bf-ce88339e3a45)

Y1 Limit signal:
![image](https://github.com/user-attachments/assets/441a073e-5acf-4756-8cb9-30f37b9400f9)
Y2 Limit Signal:
![image](https://github.com/user-attachments/assets/df4ad8be-3af0-4e23-b620-ac51239e35c6)
X: Limit Signal:
![image](https://github.com/user-attachments/assets/4f9d3e66-001a-4b8d-b187-685c9a549bc9)

Another experiment i tried disconnecting the shielding from one of the limit input wires, and to my big amazement that gave a cleaner signal on the input. So my Earth is definitely not particularly clean.

So I guess the big question is: If I have such varying noise signatures, I guess that it could also be down to the fact that i use the 5v from the USB to drive everything that uses 5v. Should i go for a 5v supply seperately?

Alternatively i was looking at the sensors, and they support up to 24v, and i already have 24v in the cabinet for driving the VFD control inputs. so I was wondering if there might be a solution in a Logic conversion board, from 24v to 5v, and then running the sensors on 24v. (could be something like this: https://www.fruugo.dk/24v-to-5v-plc-optocoupler-isolation-board-4-channel-voltage-converter-module/p-286023466-638021991?language=en&ac=croud&asc=pmax&gad_source=1&gclid=Cj0KCQjwwuG1BhCnARIsAFWBUC168UUohKjOnhwLjvqOJOsYX-QVe1O9MzrUBwmWwVJKOhHgfcskYbIaAmPkEALw_wcB) 

Or any other suggestions on what directions to look in?

**Comment by terjeio** [comment]
[This commen](https://github.com/grblHAL/core/discussions/560#discussioncomment-10264245)t implies that your switches are not NPN as advertised? If NPN and open collector the voltage at the board terminal should be be the same regardless of the switch beeing connected or not (when triggered). When not triggered it should be close to 0V (unless the output is inverted). A connected switch should never drive the voltage at the input higher than the not connected voltage.

> Alternatively i was looking at the sensors, and they support up to 24v, and i already have 24v in the cabinet for driving the VFD control inputs. so I was wondering if there might be a solution in a Logic conversion board, from 24v to 5v, and then running the sensors on 24v. 

I am not sure about the T41U5XBB board, but at least the Pro prototype I have is designed to run the switches on 12V so 12V might be a better option. Check the [manual](https://github.com/phil-barrett/grblHAL-teensy-4.x/blob/master/T41U5XBB%20v207.pdf) or perhaps @phil-barrett may chime in.

**Reply by Klippert** [reply]
Thanks for your reply,

Nope, the inputs only support 5v according to the manual. (It's actually the T41E5XBB board i've got, but i run it over the USB.)

Supposing you're right, and the endstops at not NPN but rather PNP, (it wouldn't be the first chinese manual that was wrong) - do you have a suggestion on how should I switch the wiring?

**Reply by terjeio** [reply]
I am fine reading your other comments. You wrote this:

> It's a NPN and with "white" wired to +5v it's normally closed.

Shouldn't one of the output wires be left unconnected? One is NO and the other NC?
Each output has a separate transistor - configured as open collector. If you connect the unused one to 5V you are loading the 5V line (basically shorting it to ground - depending on any current limiting resistor/circuitry present).

![355268278-26dfa776-0724-4f37-8bd7-93975690caba](https://github.com/user-attachments/assets/ad242127-15c9-4576-a5b7-84fb9663b2f4)

If the switch is really open collector you should be able to connect the  brown wire to 12V or 24V, leave the white unconnected and have 0V at the black when not connected to the board regadless of triggered status. When connected to the board you should get 0 or ~5V depending on triggered status.

**Reply by phil-barrett** [reply]
Note that the voltage you see at the signal (sig) input on the T41U5XBB will be backwards - 0V is triggered and ~4V is not triggered. You should see something close to 4 V due to the voltage drop of the LED in the optoisolator (0.9 to 1.1V). This diagram may help you to understand how things fit together.

![image](https://github.com/user-attachments/assets/9a165804-58f6-45ef-a800-8065e9274471)

**Reply by Klippert** [reply]
> Note that the voltage you see at the signal (sig) input on the T41U5XBB will be backwards - 0V is triggered and ~4V is not triggered. You should see something close to 4 V due to the voltage drop of the LED in the optoisolator (0.9 to 1.1V). This diagram may help you to understand how things fit together.
> 
> ![image](https://private-user-images.githubusercontent.com/16570958/358665403-9a165804-58f6-45ef-a800-8065e9274471.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMDkyNTEsIm5iZiI6MTcyNDAwODk1MSwicGF0aCI6Ii8xNjU3MDk1OC8zNTg2NjU0MDMtOWExNjU4MDQtNThmNi00NWVmLWE4MDAtODA2NWU5Mjc0NDcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDE5MjIzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxNjYxZjQ4ZThhYmMyNmQwY2FmNjRjMWEyY2VlMzU2ZGFkODY1M2ZjMjE2OThhOGJjMTQxNGI3ZTIzOWQ1MzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.z-XeGA8vrkY7IcgoVl2uDxzqASPq7SZGrlM2nBKnaQU)

Hello Phil, and thanks for chipping in. 

With the breakout board, would it be more robust towards EMI to use an NPN or PNP configuration on the sensor part?
And the pro board you once talked about that runs higher voltages on the sensors, did you ever put that into production?

Best
Kristoffer

**Reply by phil-barrett** [reply]
No, the Pro board is not in production. Note that the PicoCNC board is available and uses 12V limit inputs.  Works largely the same as the Teensy based board but only has 4 Axes and needs slightly different settings.  It is also a bit slower than the Teensy board but you may not see much difference at all - Lasers will see a lot of benefit though.

I don't think NPN or PNP has an advantage when it comes to EMI.

**Reply by Klippert** [reply]
Hmn. ok. Did you can the idea of the pro board, or is it just awaiting test or so. 

I do use all 5 axis outputs of the current board so 4 axis would kill my autosquaring. 
Is my understanding, that running 12 or 24v on the limit switches signals, would increase their resistance to EMI correct? 

Best and thanks for all your help.
Kristoffer

**Reply by phil-barrett** [reply]
Yes, 12 or 24V would help a lot as it takes a lot more energy to falsely trigger.

I decided not to proceed with the Pro board. There were too many issues and the cost got way out of hand. I added a number of the features to the PicoCNC board and will eventually move the Teensy version to using 12V limit inputs. I am also looking at a board using the RP 2350 (follow on to the Pico's RP 2040) that will support 5 axes.

**Reply by Klippert** [reply]
Thanks - and a shame about the pro board. I'd love to have something like that. :)

**Comment by phil-barrett** [comment]
I had not noticed this thread until Terje tagged me.  Sorry, I would have jumped on sooner had I known.

Your problem sounds like EMI, especially given the intermittent symptoms.  It certainly can be infuriating.  I'd start by looking for the source(s).  Motors, especially brushed, are a prime contributor. 

Given that using simple switches has been reliable in the past leads me to believe that your chinese sensors are fairly sensitive. Even if you have shielded cables, if the sensors are picking up the EMI then shielding won't help much. It is also possible the EMI is entering via the USB cable. You might try an experiment - wrap as much of the sensor as possible with aluminium foil and connect it to the shield.

Switching to a 5V PSU may help but if your USB cable is the EMI vector it probably will not.  Usually when you have EMI on the USB cable you get garbled communications. A typical error there is Error 1, 2 or 3 and not an Alarm. In general, I recommend Ethernet for comm rather than USB.  

[This](https://www.grbl.org/single-post/emi-science-voodoo-or-doodoo) might give you some ideas.

Phil

**Reply by Klippert** [reply]
Wow. that's a bit to digest here. :-) 

So one thing is was contemplating, was just after homing, "shorting" the Y limit switches at the electronics connectors (so the wiring inside the cabinet is still used) - and seeing if I get to a stable state or not. (Of course i the Y limit switches would be out of commission then)

Also it seems the harder I push the spindle, the faster I get a failure. Could likely be that the EMI is to do with the spindle VFD or wiring to the spindle...

**Reply by phil-barrett** [reply]
You could just turn off hard limits in Grbl settings ($21) to test that theory.

It is likely that the spindle is one of your EMI sources.  Is the VFD housed in a metal box?

**Reply by Klippert** [reply]
Ah. That's quite a bit easier, now that you mention it. If I turn off hard limits, it only uses the switches for homing, and then ignores everything that happens after that?

Yes the spindle vfd is also my prime suspect, and no, it's not in a metal case, it's just a bit "away" from the remaining electronics. 30-40cm. 
I'll test the theory, and see what happens :)

**Reply by terjeio** [reply]
Did you try with disconnecting the output wire connected to 5V?

**Reply by Klippert** [reply]
> Did you try with disconnecting the output wire connected to 5V?

yes, that was how I initially wired it up, and that made no difference. Then I read this wiring, and changed it: 
![image](https://github.com/user-attachments/assets/88ce420c-a538-43a3-8fe8-dfb880234919)

**Reply by Klippert** [reply]
> You could just turn off hard limits in Grbl settings ($21) to test that theory.
> 
> It is likely that the spindle is one of your EMI sources. Is the VFD housed in a metal box?

Ok, just did a few tests. 
First I ran a few jobs to see if I could reproduce. and only after about 1.5 hours i had the problem occur.
Then I switched $21 off, and the problem went away. I  only ran a short job, but that did run successfully.

It would seem to confirm, that the problem is EMI on the limit switches - so I'm kind of back to trying to figure out where it gets "in".
Though running without hard limits, is a nice workaround to have in the pocket ;)

Currently my shield on cables are all connected to earth from the wall socket. I saw someone connecting them to ground on the controller. Is that the way to go about it?

Best
Kristoffer

**Reply by phil-barrett** [reply]
I would connect every ground to earth, except for the stepper/servo drivers.

Do it in a star fashion.  Use a metal electronics box as the center of the star that connects directly to earth ground.  Connect all of the metal components of your machine to the metal electronics box. All your cable shields should terminate (connect) at the metal electronics box. Inside the box, connect the electronics grounds to the metal box. This gives your EMI a path to earth ground. 

I would try to figure out a way to shield the sensors.  Aluminium foil as a shield that connects to the cable shield would be an easy experiment.  By the way, I really like the metal proximity sensors because they are inside a threaded metal tube so are inherently shielded.

**Reply by Klippert** [reply]
> I would connect every ground to earth, except for the stepper/servo drivers.
> 
> Do it in a star fashion. Use a metal electronics box as the center of the star that connects directly to earth ground. Connect all of the metal components of your machine to the metal electronics box. All your cable shields should terminate (connect) at the metal electronics box. Inside the box, connect the electronics grounds to the metal box. This gives your EMI a path to earth ground.
> 
> I would try to figure out a way to shield the sensors. Aluminium foil as a shield that connects to the cable shield would be an easy experiment. By the way, I really like the metal proximity sensors because they are inside a threaded metal tube so are inherently shielded.

Ah. that's an interesting one. When i measured the difference between ground and eath, it was very noisy. But i guess offsetting that by connecting the two, is a way to go.  - the way i built my machine, the electronics sits in a wooden enclosure underneath the machine (which is a pain to service ;-) ) - so i might need to change that completely. 

Alternatively i'll make a different setup, that allows for getting the electronics into an enclosure. 

I'm also thinking to add logic converters on the inputs, to move them to 24v. 

(this is the machine in question)
![4C4FD1DD-380A-4604-8308-84C1B0DBC666_1_105_c](https://github.com/user-attachments/assets/bb0e7106-3659-4243-a323-a178d59c33a8)

**Comment by phil-barrett** [comment, marked as answer]
You might not have to completely change your electronics box.  A metal plate (Aluminium or steel) at the bottom would form a good base for the star ground to earth. Also, your VFD could be mounted on a separate metal plate (that is how mine is set up) and the plate tied to earth. A metal box is better for EMI immunity but not absolutely needed.  EMI resilience is not a binary thing - there are a series of things you can do. Sometimes the simplest works, other times you have to go the full distance.  This is why I recommend doing as much as possible when building the machine - retrofitting is always harder.

**Reply by Klippert** [reply]
Thanks Phil, 
I could make two alu based drawers without changing everything. The challenge is also that the case has braces like the end three places on the inside for strength. 

Though wiring the ground for 5v to the earth and shield in the cables sounds like the "easy" place to start?

**Reply by phil-barrett** [reply]
Yes, I'd start there. Especially since you are seeing a fair amount of noise between Gnd and Earth.

**Reply by Klippert** [reply]
Sorry for the slow response. I've had to put some other work in front. I'll get back with results as soon as I've done the experiment.
And again, thanks for all the help guys, i truely appreciate that the amazing support :-) 



Best
Kristoffer

**Reply by Klippert** [reply]
A small update. - I suddenly ran into a situation where my Y1 was triggered contiunuously. So i first looked at EMI again, I even tried powering everything  down, that wasn't the controller and the sensor itself. Same result. So i dismounted the endstop sensor, and voilas, the trigger stopped. After a bit of pushing and fiddling, I found out that when you press the sensor on the bottom, it triggers. now when it's screwed down to the machine that must have been just enough to trigger it some times. 
I'm going to look for an alternative endstop sensor. I'm pretty done with these ones and all the head aches they've given me. 
So the whole EMI situation may or may not be the coulprit - but for now i'm going to change to a different endstop. recommendations are welcome. :)

Then when i have the new endstops mounted, we will see if the EMI problem persists, or it was just a "bad" sensor type.

**Reply by phil-barrett** [reply]
I'm betting on it being the sensor.  Glad you finally found it.  I like the metal proximity type.

I did a blog on them [here](https://www.grbl.org/single-post/using-proximity-sensors-with-the-picocnc).

**Reply by Klippert** [reply]
I Certainly hope so ;) 

The amount of hours one can spend pulling ones hair over a $3 sensor being unreliable is amazing. 
However the added grounding and shielding is probably not a bad thing. :-)

I'll have a look at the proximity sensors.

**Reply by Klippert** [reply]
> I'm betting on it being the sensor. Glad you finally found it. I like the metal proximity type.
> 
> I did a blog on them [here](https://www.grbl.org/single-post/using-proximity-sensors-with-the-picocnc).

Hello @phil-barrett,  
Btw. I was reading up on the inductive sensors - and i can see that prices quickly go up :) 
I was looking at the 4 wire, M8 variants - trigger distance of 2mm - they are about $125 each. (https://www.digikey.dk/da/products/detail/pepperl-fuchs-inc/NBB2-8GM50-A0/22973244)

So I was wondering how to mount them and integrate them into the machine best. And ideally they could trigger off my aluminum side panels of the Y axis. 
![image](https://github.com/user-attachments/assets/b6242d5a-69e8-4838-a383-178f9cc5edf5)

But would they actually be able to trigger off alu, or do they need iron to detect?

**Reply by malcolmmp** [reply]
My cheapo inductive sensors will trigger on aluminum, but the aluminum has to be a lot closer than steel. On a machine with linear guides like what you've got, you should be able to adjust the sensor close enough without worry that it will strike.

For my own setup, I 3d printed some simple holders for steel screws which trigger the inductive sensors.

**Reply by Klippert** [reply]
> My cheapo inductive sensors will trigger on aluminum, but the aluminum has to be a lot closer than steel. On a machine with linear guides like what you've got, you should be able to adjust the sensor close enough without worry that it will strike.
> 
> For my own setup, I 3d printed some simple holders for steel screws which trigger the inductive sensors.

Hi Malcolm,
Sounds great, and thank for the help. Can I ask you to share a link for your sensors and a picture of your mounting for inspiration?

**Reply by malcolmmp** [reply]
Sure!

I believe the sensors I used were these:
https://www.amazon.com/dp/B0C2CDL46D

I ended up using separate opto-isolated solid-state relays along with those inductive sensors. As I recall, even though the sensors are advertised as 6-36V, they worked fine on the 5V of the grblHAL board. HOWEVER, when testing on the bench it seemed that when connected straight to the controller, there was at least one potential failure mode in which a wire could come loose and the limit wouldn't register as triggered, and the machine would then potentially overt travel. With the additional solid state relays any wire could be disconnected and grblHAL will report that limit as triggered.

I made posted a question about it at the time and followed up with my final configuration.

https://github.com/phil-barrett/grblHAL-teensy-4.x/discussions/110#discussioncomment-10131093

My machine is a little unusual, but you can get a decent idea of what's going on from these pictures. In my case I have the wiring for just about everything going to my gantry, so the sensors are all along for the ride. I figured that way I could install identical triggers at the far end to use as a max travel limit.

This is on a laser machine, so I don't have the crazy EMI of a high speed spindle to worry about. If any of the sensor wires are following the same path as the spindle wire, I'd recommend shielded cable.

For what it's worth, in the first picture the printed sensor holder is just temporary until I machine a new plate with the holder built in.

Hope that helps!



![IMG_1047](https://github.com/user-attachments/assets/bd32202e-02a5-4424-a1f1-45d0176013ed)
![IMG_1104](https://github.com/user-attachments/assets/984c8abd-370c-4d2d-a2d4-16eb2565aa0a)

**Reply by Klippert** [reply]
Hiya, Thanks,

That's a great way, to avoid the sensors ending up in the crush-zone in case of a crash. I would hate to have to replace my sensors once i get them working properly. ;) 

Now, i just need some way of sourcing the sensors in EU - but i'll have a look at that. Thanks for the pictures and the links. It's really great of you to share your experience as well as ideas.

Best 
Kristoffer

**Reply by Klippert** [reply]
ok, now i got replacement Panasonic optical endstops, and i just put them on, and they seem to work. At least as far as i've gotten today. I've also got a 24v-5v logic converter, so when i have a bit more time, i'll rewire them to run 24v for a start, as well as re wire the grounding as described above.

Then we see if things will behave reliably from here on.

Thanks again for all the help everybody. 

Best
Kristoffer

**Reply by Klippert** [reply]
Just wanted to give an update. Since replacing the sensors (all of them) with not so cheep panasonic sensors with same specs, i've not had any problems. 

I've not gotten around to install the 24v converter board or anything else fancy, just milled away at different projects :)

So I consider the case closed. Just wish I'd thought of testing the sensors when tightened to different degrees earlier, but hey. if anyone else are using cheap Chinese optical sensors, do test that they read consistently when screwed firmly in place.

Best
Kristoffer

**Reply by phil-barrett** [reply]
Very glad to hear that you solved the problem. The pain of the extra cost will soon be forgotten as you run jobs that don't fail.

Could you post a link to the panasonic sensors? It might help some future user to avoid the problems you went through.

Phil

**Reply by Klippert** [reply]
Hello,

Certainly. Happy to share:
https://www.digikey.dk/da/products/detail/panasonic-industrial-automation-sales/PM-L45/5962571?so=88482037&content=productdetail_DK&mkt_tok=MDI4LVNYSy01MDcAAAGVQVzq9j4MVl12xPpt19AGwCdd7qdCqnX1sQYOVA4856iFP4jVIWDeonxfDy4OrxAhCCyeBqYhyPZXA22c36m9wXCkSwbYW59MkCwzt0Of

This is also where i bought it, but i'm sure you can find the same in other shops. Probably also cheaper.

Hope this benefits other people, 

Best
Kristoffer
